go/vt/vttablet/tabletserver/messager/message_manager.go

```diff
@@ -173,61 +174,78 @@ type messageManager struct {
 	batchSize    int
 	pollerTicks  *timer.Timer
 	purgeTicks   *timer.Timer
-	conns        *connpool.Pool
 	postponeSema *sync2.Semaphore
 
 	mu     sync.Mutex
 	isOpen bool
-	// cond gets triggered if a receiver becomes available (curReceiver != -1),
-	// an item gets added to the cache, or if the manager is closed.
-	// The trigger wakes up the runSend thread.
+	// cond waits on curReceiver == -1 || cache.IsEmpty():
+	// No current receivers available or cache is empty.
 	cond            sync.Cond
 	cache           *cache
 	receivers       []*receiverWithStatus
 	curReceiver     int
 	messagesPending bool
 
+	// streamMu keeps the cache and database consistent with each other.
+	// Specifically:
+	// It prevents items from being removed from cache while the poller
+	// reads from the db and adds items to it. Otherwise, the poller
+	// might add an older snapshot of a row that was just postponed.
+	// It blocks vstream from receiving messages while the poller
+	// reads a snapshot and updates lastPollPosition. Any events older than
+	// lastPollPosition must be ignored by the vstream. It consequently
+	// also blocks vstream from updating the cache while the poller is
+	// active.
+	streamMu         sync.Mutex
+	streamCancel     func()
```

Comment:
In startVStream/stopVStream, it appears like you are also using streamCancel as a sentinel value for the state of the stream, allowing you to idempotently start/stop streams. If I'm reading that correctly, I think it's worth adding a comment here.